cli.js

```diff
@@ -72,101 +92,81 @@ if (platform !== 'javascript') {
   }
 
-  Promise.all([...inputFilePromises, ...externFilePromises])
-    .then(files => {
-      const inputFiles = files.slice(0, inputFilePromises.length);
-      const externs = files.slice(inputFilePromises.length);
-      if (externs.length > 0) {
-        compilerFlags.externs = externs;
-      } else {
-        delete compilerFlags.externs;
+  const startCompile = !waitOnStdIn ? Promise.resolve([]) : new Promise(resolve => {
+    let stdInData = '';
+    const waitingTimeout = setTimeout(() => {
+      process.stderr.write('The compiler is waiting for input via stdin.\n');
+    }, 1000);
+    process.stdin.setEncoding('utf8');
+    process.stdin.on('readable', () => {
+      const chunk = process.stdin.read();
+      if (chunk !== null) {
+        stdInData += chunk;
+        clearTimeout(waitingTimeout);
       }
-
-      if (!waitOnStdIn) {
-        return inputFiles;
+    });
+    process.stdin.on('error', (err) => {
+      process.exitCode = 1;
+      console.error(err);
+      clearTimeout(waitingTimeout);
+    });
+    process.stdin.on('end', () => {
+      if (stdInData.length > 0) {
+        resolve([{
+          path: 'stdin',
+          src: stdInData
+        }]);
       } else {
-        return new Promise(resolve => {
-          let stdInData = '';
-          process.stdin.setEncoding('utf8');
-          process.stdin.on('readable', () => {
-            const chunk = process.stdin.read();
-            if (chunk !== null) {
-              stdInData += chunk;
-            }
-          });
-          process.stdin.on('error', (err) => {
-            process.exitCode = 1;
-            console.error(err);
-          });
-          process.stdin.on('end', () => {
-            if (stdInData.length > 0) {
-              inputFiles.push({
-                path: 'stdin',
-                src: stdInData
-              });
-            }
-            resolve(inputFiles);
-          });
-        });
+        resolve([]);
       }
-    })
-    .then(inputFiles => {
+      clearTimeout(waitingTimeout);
+    });
+  });
+
+  startCompile.then(inputFiles => {
       const Compiler = require('./lib/node/closure-compiler-js');
       const logErrors = require('./lib/logger');
       const compiler = new Compiler(compilerFlags);
       const output = compiler.run(inputFiles);
-      const exitCode = output.errors.length === 0 ? 0 : 1;
+      if (output.errors.length > 0) {
+        process.exitCode = process.exitCode || 1;
+      }
       logErrors(output, inputFiles);
-      if (output.compiledFiles.length === 1 && output.compiledFiles[0].path === 'compiled.js' &&
-          !compilerFlags['js_output_file']) {
-        console.log(output.compiledFiles[0].src);
+      if (output.compiledFiles.length > 0) {
+        if (compilerFlags['js_output_file'] || compilerFlags['chunk']) {
+          let srcMapPattern = '%outname%.map';
+          if (compilerFlags.create_source_map && typeof compilerFlags.create_source_map === 'string') {
```

Comment:
Could you add some comments on these if conditions?
I find them less than obvious.
For example, I'm guessing that `--create_source_map` or `--create_source_map=true` creates a boolean value, but any string other than true/false is assumed to be a path for the map. Is that what's happening here?